pkg/sigv4/utils.go

```diff
@@ -0,0 +1,98 @@
+package sigv4
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+var noEscape [256]bool
+
+// escapePath escapes part of a URL path in Amazon style.
+func escapePath(path string, encodeSep bool) string {
+	var buf bytes.Buffer
+	for i := 0; i < len(path); i++ {
+		c := path[i]
+		if noEscape[c] || (c == '/' && !encodeSep) {
+			buf.WriteByte(c)
+		} else {
+			fmt.Fprintf(&buf, "%%%02X", c)
+		}
+	}
+	return buf.String()
+}
+
+// stripExcessSpaces will rewrite the passed in slice's string values to not
+// contain multiple side-by-side spaces.
+func stripExcessSpaces(str string) string {
```

Comment:
This method (and maybe others from this file) looks utterly identical to the one from AWS SDK. Unfortunately, I'm not sure if we can accept it as it is because of licensing. Its performance is also less of a concern here. Can we rewrite it (and other methods if needed) by hand?

Reply:
Yes these ones where a ported straight for the sdk. I can try to rewrite it by hand. I will look into the test cases for it too and see If I can get them to pass with the new rewrite.

Reply:
There are couple more here that ported and made slight changes, would you like me to look into rewriting them?

Reply:
I mean, the rule of thumb is that we should not copy code directly, so yeah, if rewriting it on your own is preferable. But since, in any case, this PR is aiming to re-implement some logic to make the output AWS compatible, which still forces us that or another way use similar approaches so I believe at some point, we should still left some credits to AWS SDK with something like `logic is highly inspired by the AWS SDK` or any better referencing

Reply:
Yes thats exactly it, I initially didn't want to rewrite it since I wanted the exact aws implementation. My thinking is we want to use the SDK behaviour, but without including it, so I figured porting it would be less risk of mistakes.

So are you okay if I add a comment with attributions instead of rolling a new one?

Reply:
No, that's not what I meant, sorry :cry: I was saying that in all cases, we need to mention that logic of that package tries to mimic AWS SDK. But having an utterly identical method, even if we reference to AWS SDK, isn't way to go